extensions/functions_datetime.yaml

```diff
@@ -265,3 +354,264 @@ scalar_functions:
+            value: i64
+          - name: origin
+            value: time
+        return: time
+  -
+    name: "round_calendar"
+    description: >-
+      Round a given timestamp/date/time to a multiple of a time unit. Time unit can be physical or calendar based.
+      Input value is floored to multiple of temporal units until the next origin unit. Conversely value would be
+      ceiled to the multiple of temporal units since the last origin unit. Rounding would return the closest of the
+      ceiled and floored results.
+
+      Timezone strings must be as defined by IANA timezone database (https://www.iana.org/time-zones).
+      Examples: "Pacific/Marquesas", "Etc/GMT+1".
+      If timezone is not present then UTC is assumed. If timezone is invalid an error is thrown.
+    impls:
+      - args:
+          - name: x
+            value: timestamp
+          - name: rounding
+            options: [ FLOOR, CEIL, ROUND ]
+            required: true
+          - name: unit
+            options: [ YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND, MICROSECOND ]
+            required: true
+          - name: origin
+            options: [ YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND ]
+            required: true
+          - name: multiple
+            value: i64
+        return: timestamp
+      - args:
+          - name: x
+            value: timestamp_tz
+          - name: rounding
+            options: [ FLOOR, CEIL, ROUND ]
+            required: true
+          - name: unit
+            options: [ YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND, MICROSECOND ]
+            required: true
+          - name: origin
+            options: [ YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND ]
+            required: true
+          - name: multiple
+            value: i64
+          - name: timezone
+            description: Timezone string from IANA tzdb.
+            value: string
+        return: timestamp_tz
+      - args:
+          - name: x
+            value: date
+          - name: rounding
+            options: [ FLOOR, CEIL, ROUND ]
+            required: true
+          - name: unit
+            options: [ YEAR, MONTH, WEEK, DAY ]
+            required: true
+          - name: origin
+            options: [ YEAR, MONTH, WEEK, DAY ]
+            required: true
+          - name: multiple
+            value: i64
+          - name: origin
+            value: date
+        return: date
+      - args:
+          - name: x
+            value: time
+          - name: rounding
+            options: [ FLOOR, CEIL, ROUND ]
+          - name: unit
+            options: [ DAY, HOUR, MINUTE, SECOND, MILLISECOND, MICROSECOND ]
+            required: true
+          - name: origin
+            options: [ DAY, HOUR, MINUTE, SECOND, MILLISECOND ]
+            required: true
+          - name: multiple
+            value: i64
+          - name: origin
+            value: time
+        return: time
```

Comment:
Rounding to unit-based-origin is [lubridate](https://lubridate.tidyverse.org/reference/round_date.html) behaviour and as per experiences in the [Acero rounding kernel](https://github.com/apache/arrow/pull/12657#issuecomment-1119580484) it is best separated from the instant-based-origin.

Reply:
This behavior comes across as really confusing to me, but I suppose it does seem to cover the edge cases.

Reply:
It's useful if you want to round to say 3 days, but not from 1970-01-01 but rather from the beginning of each month.
Rounding timestamps (in local time) is a deep rabbit hole.

Reply:
I understand what it's useful for, it's mostly just the combination of floor/ceil and since/until in that description that's confusing me. So I ended up working out manually what your description actually implies in terms of behavior when taken to the letter.

> Input value is floored to multiple of temporal units until the next origin unit.

-> Round input value *down* such that there is an integer number of `multiple`s between it and the next `origin`.

though it could also be interpreted as

-> Round input value *up* such that there is an integer number of `multiple`s between it and the next `origin`.

if "floored" refers to the *difference* between the input and the next origin unit, rather than the input itself. I'll call those "floored value" and "floored difference". Based on the rounding options argument I guess the correct one is the one that actually floors the value, but it makes a huge mess for the rounded version.

> Conversely value would be ceiled to the multiple of temporal units since the last origin unit.

-> Round input value *up* such that there is an integer number of `multiple`s between it and the previous `origin`.

This is not ambiguous because there's no negation here due to the "time until" and always ceils. However, the way it's written implies to me that it both flooring and ceiling happen at the same and that an assertion is made that this is the same thing or something. It'd be very helpful to actually refer to the options here, i.e. "Conversely, *if rounding is set to `CEILING`*, value would be ceiled...".

> Rounding would return the closest of the ceiled and floored results.

-> Return whichever of the above two values is closer to the input.

I think this is actually ambiguous though. What if the input is equidistant to both options?

I ended up just writing my various algorithm interpretations as a Python script to try it out, because I got way confused. The script operates on integers rather than calendar units, but the logic is the same. I used multiple = 300 and origin = 1000 (so that could be rounding to multiples of 300 milliseconds with the origin set to whole seconds) to hit all corner cases due to indivisibility and halfway values.

<details><summary>Results</summary>

| Input | Floored (val) | Floored (dif) | Ceiled | Rounded (val) | Rounded (dif) |
|-------|---------------|---------------|--------|---------------|---------------|
|  -100 |          -300 |             0 |   -100 |          -100 |          -100 |
|   -50 |          -300 |             0 |    200 |   -300 or 200 |             0 |
|     0 |             0 |             0 |      0 |             0 |             0 |
|    50 |          -200 |           100 |    300 |   -200 or 300 |           100 |
|   100 |           100 |           100 |    300 |           100 |           100 |
|   150 |           100 |           400 |    300 |           100 |           300 |
|   200 |           100 |           400 |    300 |    100 or 300 |           300 |
|   250 |           100 |           400 |    300 |           300 |           300 |
|   300 |           100 |           400 |    300 |           300 |           300 |
|   350 |           100 |           400 |    600 |    100 or 600 |           400 |
|   400 |           400 |           400 |    600 |           400 |           400 |
|   450 |           400 |           700 |    600 |           400 |           600 |
|   500 |           400 |           700 |    600 |    400 or 600 |           600 |
|   550 |           400 |           700 |    600 |           600 |           600 |
|   600 |           400 |           700 |    600 |           600 |           600 |
|   650 |           400 |           700 |    900 |    400 or 900 |           700 |
|   700 |           700 |           700 |    900 |           700 |           700 |
|   750 |           700 |          1000 |    900 |           700 |           900 |
|   800 |           700 |          1000 |    900 |    700 or 900 |           900 |
|   850 |           700 |          1000 |    900 |           900 |           900 |
|   900 |           700 |          1000 |    900 |           900 |           900 |
|   950 |           700 |          1000 |   1200 |   700 or 1200 |          1000 |
|  1000 |          1000 |          1000 |   1000 |          1000 |          1000 |
|  1050 |           800 |          1100 |   1300 |   800 or 1300 |          1100 |
|  1100 |          1100 |          1100 |   1300 |          1100 |          1100 |


```python
import math

def ceil_multiple(input, multiple):
    """Round integer up to multiple from 0."""
    return (input + multiple - 1) // multiple * multiple

def floor_multiple(input, multiple):
    """Round integer down to multiple from 0."""
    return input // multiple * multiple

def closest(input, a, b):
    """Pick a or b depending on which is closer to input. Return result as a
    string. If a and b are different and equidistant from input, return the string
    "{a} or {b}"."""
    if a == b:
        return f"{a}"
    da = abs(input - a)
    db = abs(input - b)
    if da < db:
        return f"{a}"
    elif db < da:
        return f"{b}"
    else:
        return f"{a} or {b}"

def round_calendar(input, multiple, origin):
    """Implement the proposed rounding logic. Rather than returning the result,
    print a table line with the intermediate results."""
    # "Input value is floored to multiple of temporal units until the next origin unit."
    next_origin = ceil_multiple(input, origin)

    # Both interpretations:
    floored_value = floor_multiple(input - next_origin, multiple) + next_origin
    floored_difference = next_origin - floor_multiple(next_origin - input, multiple)

    # "Conversely value would be ceiled to the multiple of temporal units since the last origin unit."
    previous_origin = floor_multiple(input, origin)
    ceiled = ceil_multiple(input - previous_origin, multiple) + previous_origin

    # "Rounding would return the closest of the ceiled and floored results."
    rounded_value = closest(input, floored_value, ceiled)
    rounded_difference = closest(input, floored_difference, ceiled)

    print(f"| {input:>5} | {floored_value:>13} | {floored_difference:>13} | {ceiled:>6} | {rounded_value:>13} | {rounded_difference:>13} |")

multiple = 300
origin = 1000

print("| Input | Floored (val) | Floored (dif) | Ceiled | Rounded (val) | Rounded (dif) |")
print("|-------|---------------|---------------|--------|---------------|---------------|")
for input in range(-100, 1101, 50):
    round_calendar(input, multiple, origin)
```

</details>

If we're using the interpretation of the floored version that actually floored, the rounding version makes a completely indecipherable mess, so I'm pretty sure there's something fishy about the way the operation is currently described...

Reply:
I've thought about this some more and this is how I would write/implement this instead:

> Round a given timestamp/date/time to a multiple of a time unit. If the given timestamp is not already an exact multiple of `origin` in the given `timezone`, the origin point is chosen as one of the two nearest multiples. Which of these is chosen is governed by `origin_rounding`: `FLOOR` means to use the earlier one, `CEIL` means to use the later one, `ROUND_TIE_DOWN` means to choose the nearest and tie to the earlier one if equidistant, `ROUND_TIE_UP` means to choose the nearest and tie to the later one if equidistant. If the given timestamp is not already an exact multiple of `multiple` `unit`s away from the chosen origin point, the function chooses between one of the two nearest multiples. Which of these is chosen is governed by `rounding`, using the same logic as `origin_rounding`.

<details><summary>Results</summary>

For equal `rounding` and `origin_rounding`:

| Input | Floor | Ceil | Round-tie-up | Round-tie-down |
|-------|-------|------|--------------|----------------|
|  -100 |  -100 |    0 |            0 |              0 |
|   -50 |  -100 |    0 |            0 |              0 |
|     0 |     0 |    0 |            0 |              0 |
|    50 |     0 |  100 |            0 |              0 |
|   100 |     0 |  100 |            0 |              0 |
|   150 |     0 |  400 |          300 |              0 |
|   200 |     0 |  400 |          300 |            300 |
|   250 |     0 |  400 |          300 |            300 |
|   300 |   300 |  400 |          300 |            300 |
|   350 |   300 |  400 |          300 |            300 |
|   400 |   300 |  400 |          300 |            300 |
|   450 |   300 |  700 |          600 |            300 |
|   500 |   300 |  700 |          400 |            600 |
|   550 |   300 |  700 |          700 |            400 |
|   600 |   600 |  700 |          700 |            700 |
|   650 |   600 |  700 |          700 |            700 |
|   700 |   600 |  700 |          700 |            700 |
|   750 |   600 | 1000 |          700 |            700 |
|   800 |   600 | 1000 |          700 |            700 |
|   850 |   600 | 1000 |         1000 |            700 |
|   900 |   900 | 1000 |         1000 |           1000 |
|   950 |   900 | 1000 |         1000 |           1000 |
|  1000 |  1000 | 1000 |         1000 |           1000 |
|  1050 |  1000 | 1100 |         1000 |           1000 |
|  1100 |  1000 | 1100 |         1000 |           1000 |

```python
import math

def floor_multiple(input, multiple):
    """Round integer down to multiple from 0."""
    return input // multiple * multiple

def ceil_multiple(input, multiple):
    """Round integer up to multiple from 0."""
    return (input + multiple - 1) // multiple * multiple

def round_tie_down_multiple(input, multiple):
    """Round integer to the nearest multiple from 0, tie down if equidistant."""
    return (input + (multiple - 1) // 2) // multiple * multiple

def round_tie_up_multiple(input, multiple):
    """Round integer to the nearest multiple from 0, tie down if equidistant."""
    return (input + multiple // 2) // multiple * multiple

def round_multiple(input, multiple, method):
    return {
        "FLOOR": floor_multiple,
        "CEIL": ceil_multiple,
        "ROUND_TIE_DOWN": round_tie_down_multiple,
        "ROUND_TIE_UP": round_tie_up_multiple,
    }[method](input, multiple)

def round_calendar(input, multiple, origin, rounding, origin_rounding):
    """Implement the proposed rounding logic."""
    # "If the given timestamp is not already an exact multiple of `origin` in
    # the given `timezone`, the origin point is chosen as one of the two
    # nearest multiples."
    origin = round_multiple(input, origin, origin_rounding)

    # "If the given timestamp is not already an exact multiple of `multiple`
    # `unit`s away from the chosen origin point, the function chooses between
    # one of the two nearest multiples."
    return round_multiple(input - origin, multiple, rounding) + origin

multiple = 300
origin = 1000

print("| Input | Floor | Ceil | Round-tie-up | Round-tie-down |")
print("|-------|-------|------|--------------|----------------|")
for input in range(-100, 1101, 50):
    floored = round_calendar(input, multiple, origin, "FLOOR", "FLOOR")
    ceiled = round_calendar(input, multiple, origin, "CEIL", "CEIL")
    rounded_tied_up = round_calendar(input, multiple, origin, "ROUND_TIE_UP", "ROUND_TIE_UP")
    rounded_tied_down = round_calendar(input, multiple, origin, "ROUND_TIE_DOWN", "ROUND_TIE_DOWN")
    print(f"| {input:>5} | {floored:>5} | {ceiled:>4} | {rounded_tied_up:>12} | {rounded_tied_down:>14} |")
```

</details>

There are still some shenanigans when the origin is rounded (i.e. the crossover point is halfway), but I think that's just the nature of the beast when used in conjunction with a multiple that doesn't divide the origin time unit...?

This is indeed a rabbit hole, but if we want to adopt this function in the spec, we'll have to go down it nonetheless. Might be worth splitting it off into another PR if this drags on.

Reply:
Yeah, that makes sense, so how about:
```suggestion
    name: "round_calendar"
    description: >-
      Round a given timestamp/date/time to a multiple of a time unit. If the given timestamp is not already an exact
      multiple from origin in the given timezone, the resulting point is chosen as one of the two nearest multiples.
      Which of these is chosen is governed by origin_rounding: FLOOR means to use the earlier one, CEIL means to 
      use the later one, ROUND_TIE_DOWN means to choose the nearest and tie to the earlier one if equidistant,
      ROUND_TIE_UP means to choose the nearest and tie to the later one if equidistant. If the given timestamp is
      not already an exact multiple of multiple units away from the chosen origin point, the function chooses
      between one of the two nearest multiples. Which of these is chosen is governed by rounding, using the same
      logic as origin_rounding.

      Timezone strings must be as defined by IANA timezone database (https://www.iana.org/time-zones).
      Examples: "Pacific/Marquesas", "Etc/GMT+1".
      If timezone is not present then UTC is assumed. If timezone is invalid an error is thrown.
      
    impls:
      - args:
          - name: x
            value: timestamp
          - name: rounding
            options: [ FLOOR, CEIL, ROUND_TIE_DOWN,  ROUND_TIE_UP ]
            required: true
          - name: unit
            options: [ YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND, MICROSECOND ]
            required: true
          - name: origin
            options: [ YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND ]
            required: true
          - name: multiple
            value: i64
        return: timestamp
      - args:
          - name: x
            value: timestamp_tz
          - name: rounding
            options: [ FLOOR, CEIL, ROUND_TIE_DOWN,  ROUND_TIE_UP ]
            required: true
          - name: unit
            options: [ YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND, MICROSECOND ]
            required: true
          - name: origin
            options: [ YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND ]
            required: true
          - name: multiple
            value: i64
          - name: timezone
            description: Timezone string from IANA tzdb.
            value: string
        return: timestamp_tz
      - args:
          - name: x
            value: date
          - name: rounding
            options: [ FLOOR, CEIL, ROUND_TIE_DOWN,  ROUND_TIE_UP ]
            required: true
          - name: unit
            options: [ YEAR, MONTH, WEEK, DAY ]
            required: true
          - name: origin
            options: [ YEAR, MONTH, WEEK, DAY ]
            required: true
          - name: multiple
            value: i64
          - name: origin
            value: date
        return: date
      - args:
          - name: x
            value: time
          - name: rounding
            options: [ FLOOR, CEIL, ROUND_TIE_DOWN,  ROUND_TIE_UP ]
          - name: unit
            options: [ DAY, HOUR, MINUTE, SECOND, MILLISECOND, MICROSECOND ]
            required: true
          - name: origin
            options: [ DAY, HOUR, MINUTE, SECOND, MILLISECOND ]
            required: true
          - name: multiple
            value: i64
          - name: origin
            value: time
        return: time
```

Reply:
Also if you'd like to use an implementation see:
https://arrow.apache.org/docs/python/generated/pyarrow.compute.floor_temporal.html
```python
import pyarrow as pa
from pyarrow import compute as pc

arr = pa.array([10**14, 10**15, 10**16], type=pa.timestamp("us", "Pacific/Marquesas"))
pc.floor_temporal(arr, 3, "hour", calendar_based_origin=True)
```
```
<pyarrow.lib.TimestampArray object at 0x11fb4bd60>
[
  1973-03-03 09:30:00.000000,
  2001-09-09 00:30:00.000000,
  2286-11-20 15:30:00.000000
]
```

Reply:
For my definition there would be two rounding options;

```
          - name: rounding
            options: [ FLOOR, CEIL, ROUND_TIE_DOWN, ROUND_TIE_UP ]
          - name: origin_rounding
            options: [ FLOOR, CEIL, ROUND_TIE_DOWN, ROUND_TIE_UP ]
```

(there's a double space before `ROUND_TIE_UP` in your suggestion, by the way; the YAML linter/format check would probably complain about that)

Using the same option for both would also work, kinda, but they do fundamentally different things; one chooses the crossover point for which origin to use (where shenanigans might occur) and the other chooses how to actually round toward that point.

Reply:
In my mind origin in `"round_calendar"` is always before the given timestamp. And by origin I imagine like 0 in integers. Is there a benefit to choosing the origin?

Reply:
> In my mind origin in "round_calendar" is always before the given timestamp.

What I'm pretty sure you meant to say corresponds to origin_rounding = FLOOR by my definition, but strictly speaking that definition gets pretty weird when the incoming value aligns exactly with the origin point.

> Is there a benefit to choosing the origin?

I don't know; it seemed to generalize nicely that way, but I'm not attached to it. If everyone uses `FLOOR` behavior for `origin_rounding` we can leave the option out and update the description accordingly.

Reply:
> You probably meant "before or equal to".

Yes.

> I don't know; it seemed to generalize nicely that way, but I'm not attached to it. If everyone uses FLOOR behavior for origin_rounding we can leave the option out and update the description accordingly.

I suppose it's nicely symmetrical but I'd avoid as I've not seen it elsewhere. I'll update the description.